server.go

```diff
@@ -4174,13 +4325,303 @@ func (s *server) SubscribeCustomMessages() (*subscribe.Client, error) {
+			// slot, so we can demote this peer.
+			s.peerScores[remoteHex] = peerSlotStatus{
+				state: peerStatusRestricted,
+			}
+
+			// Update numRestricted.
+			s.numRestricted++
+
+			s.banScoreMtx.Unlock()
+
+			return nil
+		}
+
+		// Else, we don't need to demote this peer since it has other
+		// pending-open channels with us.
+		peerCount.PendingOpenCount = currentNumPending
+		s.peerCounts[remoteHex] = peerCount
+
+		s.banScoreMtx.Unlock()
+
+		return nil
+
+	case peerStatusRestricted:
+		s.banScoreMtx.Unlock()
+
+		// This should not be possible. This indicates an error.
+		return fmt.Errorf("invalid peer access state transition")
+
+	default:
+		s.banScoreMtx.Unlock()
+
+		// This should not be possible.
+		return fmt.Errorf("invalid peer access status")
+	}
+}
+
+// NewOpenChan is called when a pending-open channel becomes an open channel
+// (i.e. the funding transaction has confirmed). If the remote peer is a
+// temporary-access peer, it will be promoted to a protected-access peer.
+func (s *server) NewOpenChan(remoteHex string) error {
+	s.banScoreMtx.Lock()
+	defer s.banScoreMtx.Unlock()
+
+	// Fetch the peer's access status from peerScores.
+	status, found := s.peerScores[remoteHex]
+	if !found {
+		// If we didn't find the peer, we'll return an error.
+		return ErrNoPeerScore
+	}
+
+	switch status.state {
+	case peerStatusProtected:
+		// If the peer's state is already protected, we don't need to
+		// do anything more.
+		return nil
+
+	case peerStatusTemporary:
+		// If the peer's state is temporary, we'll upgrade the peer to
+		// a protected peer.
+		peerCount, found := s.peerCounts[remoteHex]
+		if !found {
+			// Error if we did not find any info in peerCounts.
+			return ErrNoPendingPeerInfo
+		}
+
+		peerCount.HasOpenOrClosedChan = true
+		s.peerCounts[remoteHex] = peerCount
+
+		newStatus := peerSlotStatus{
+			state: peerStatusProtected,
+		}
+		s.peerScores[remoteHex] = newStatus
+
+		return nil
+
+	case peerStatusRestricted:
+		// This should not be possible. For the server to receive a
+		// state-transition event via NewOpenChan, the server must have
+		// previously granted this peer "temporary" access. This
+		// temporary access would not have been revoked or downgraded
+		// without `CloseChannel` being called with the pending
+		// argument set to true. This means that an open-channel state
+		// transition would be impossible. Therefore, we can return an
+		// error.
+		return fmt.Errorf("invalid peer access status")
+
+	default:
+		// This should not be possible.
+		return fmt.Errorf("invalid peer access status")
+	}
+}
+
+// checkIncomingConnBanScore checks whether, given the remote's public hex-
+// encoded key, we should accept this incoming connection or immediately
+// disconnect. This does not assign to the server's peerScores maps. This is
+// just an inbound filter that the brontide listeners use.
+func (s *server) checkIncomingConnBanScore(remoteHex string) (bool, error) {
```

Comment:
hmm maybe call it `checkProofOfChannelScore` because it does only check for one of our metrics ?

Reply:
I'd rather not change the name here unless it's really confusing